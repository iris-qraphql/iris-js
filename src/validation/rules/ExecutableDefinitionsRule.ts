import { GraphQLError } from '../../error/GraphQLError';

import { isExecutableDefinitionNode } from '../../language/predicates';
import type { ASTVisitor } from '../../language/visitor';

import type { ASTValidationContext } from '../ValidationContext';

/**
 * Executable definitions
 *
 * A GraphQL document is only valid for execution if all definitions are either
 * operation or fragment definitions.
 *
 * See https://spec.graphql.org/draft/#sec-Executable-Definitions
 */
export function ExecutableDefinitionsRule(
  context: ASTValidationContext,
): ASTVisitor {
  return {
    Document(node) {
      for (const definition of node.definitions) {
        if (!isExecutableDefinitionNode(definition)) {
          context.reportError(
            new GraphQLError(
              `The "${definition.name.value}" definition is not executable.`,
              definition,
            ),
          );
        }
      }
      return false;
    },
  };
}
