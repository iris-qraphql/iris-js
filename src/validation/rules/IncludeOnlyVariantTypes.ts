import { Kind } from 'graphql';

import type { ASTVisitor } from '../../language/visitor';

import { irisNodeError } from '../../error';
import type {
  TypeDefinitionNode,
  VariantDefinitionNode,
} from '../../types/ast';
import { scalarNames } from '../../types/definition';
import { IrisKind } from '../../types/kinds';

import type { SDLValidationContext } from '../ValidationContext';

const makeScalar = (name: string): TypeDefinitionNode => ({
  kind: IrisKind.TYPE_DEFINITION,
  role: 'data',
  name: { kind: Kind.NAME, value: name },
  variants: [],
});

const defaultTypes: Record<string, TypeDefinitionNode> = Object.fromEntries(
  scalarNames.map((name) => [name, makeScalar(name)]),
);

export function IncludeOnlyVariantTypes(
  context: SDLValidationContext,
): ASTVisitor {
  const doc = context.getDocument().definitions;

  return {
    TypeDefinition: checkVariantUniqueness,
  };

  function checkVariantUniqueness(type: TypeDefinitionNode) {
    const typeName = type.name.value;

    type.variants.forEach((v: VariantDefinitionNode) => {
      const name = v.name.value;
      const member =
        !v.fields &&
        (defaultTypes[name] ?? doc.find((x) => x.name.value === name));

      if (!member) {
        return undefined;
      }

      if (
        member.kind !== IrisKind.TYPE_DEFINITION ||
        member.role !== type.role ||
        !isTypeVariantNode(member)
      ) {
        context.reportError(
          irisNodeError(
            `${type.role} ${typeName} can only include ${type.role} variantTypes, it cannot include ${member.role} ${member.name.value}.`,
            v,
          ),
        );
      }
    });

    return false;
  }
}

const isTypeVariantNode = (type: TypeDefinitionNode) => {
  const { variants } = type;

  if (variants.length === 0) {
    return true;
  }

  const [variant] = variants;
  const typeName = type.name.value;

  return (
    variants.length === 1 &&
    variant.name.value === typeName &&
    variant.fields !== undefined
  );
};
