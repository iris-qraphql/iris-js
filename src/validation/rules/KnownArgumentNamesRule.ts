import { IrisKind } from '../../language/kinds';
import type { ASTVisitor } from '../../language/visitor';

import { specifiedDirectives } from '../../type/directives';

import { irisError } from '../../error';
import { didYouMean, suggestionList } from '../../utils/legacy';

import type { SDLValidationContext } from '../ValidationContext';

/**
 * @internal
 */
export function KnownArgumentNamesOnDirectivesRule(
  context: SDLValidationContext,
): ASTVisitor {
  const directiveArgs = Object.create(null);

  const schema = context.getSchema();
  const definedDirectives = schema ? schema.directives : specifiedDirectives;
  for (const directive of definedDirectives) {
    directiveArgs[directive.name] = directive.args.map((arg) => arg.name);
  }

  const astDefinitions = context.getDocument().definitions;
  for (const def of astDefinitions) {
    if (def.kind === IrisKind.DIRECTIVE_DEFINITION) {
      const argsNodes = def.arguments ?? [];

      directiveArgs[def.name.value] = argsNodes.map((arg) => arg.name.value);
    }
  }

  return {
    Directive(directiveNode) {
      const directiveName = directiveNode.name.value;
      const knownArgs = directiveArgs[directiveName];

      if (directiveNode.arguments && knownArgs) {
        for (const argNode of directiveNode.arguments) {
          const argName = argNode.name.value;
          if (!knownArgs.includes(argName)) {
            const suggestions = suggestionList(argName, knownArgs);
            context.reportError(
              irisError(
                `Unknown argument "${argName}" on directive "@${directiveName}".` +
                  didYouMean(suggestions),
                { node: argNode },
              ),
            );
          }
        }
      }

      return false;
    },
  };
}
