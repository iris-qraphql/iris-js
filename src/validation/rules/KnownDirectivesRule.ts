import type { Kind } from 'graphql';

import type { ASTNode } from '../../language/ast';
import { DirectiveLocation } from '../../language/directiveLocation';
import { IrisKind } from '../../language/kinds';
import type { ASTVisitor } from '../../language/visitor';

import { specifiedDirectives } from '../../type/directives';

import { irisError } from '../../error';
import { inspect, invariant } from '../../utils/legacy';

import type {
  SDLValidationContext,
  ValidationContext,
} from '../ValidationContext';

/**
 * Known directives
 *
 * A GraphQL document is only valid if all `@directives` are known by the
 * schema and legally positioned.
 *
 * See https://spec.graphql.org/draft/#sec-Directives-Are-Defined
 */
export function KnownDirectivesRule(
  context: ValidationContext | SDLValidationContext,
): ASTVisitor {
  const locationsMap = Object.create(null);

  const schema = context.getSchema();
  const definedDirectives = schema
    ? schema.getDirectives()
    : specifiedDirectives;
  for (const directive of definedDirectives) {
    locationsMap[directive.name] = directive.locations;
  }

  const astDefinitions = context.getDocument().definitions;
  for (const def of astDefinitions) {
    if (def.kind === IrisKind.DIRECTIVE_DEFINITION) {
      locationsMap[def.name.value] = def.locations.map((name) => name.value);
    }
  }

  return {
    Directive(node, _key, _parent, _path, ancestors) {
      const name = node.name.value;
      const locations = locationsMap[name];

      if (!locations) {
        context.reportError(
          irisError(`Unknown directive "@${name}".`, { node }),
        );
        return;
      }

      const candidateLocation = getDirectiveLocationForASTPath(ancestors);
      if (candidateLocation && !locations.includes(candidateLocation)) {
        context.reportError(
          irisError(
            `Directive "@${name}" may not be used on ${candidateLocation}.`,
            { node },
          ),
        );
      }
    },
  };
}

function getDirectiveLocationForASTPath(
  ancestors: ReadonlyArray<ASTNode | ReadonlyArray<ASTNode>>,
): DirectiveLocation | undefined {
  const appliedTo = ancestors[ancestors.length - 1];
  invariant('kind' in appliedTo);

  switch (appliedTo.kind) {
    case IrisKind.FIELD_DEFINITION:
      return DirectiveLocation.FIELD_DEFINITION;
    case IrisKind.TYPE_DEFINITION:
      return DirectiveLocation.UNION;
    case IrisKind.ARGUMENT_DEFINITION: {
      const parentNode = ancestors[ancestors.length - 3];
      invariant('kind' in parentNode);
      const kinds: ReadonlyArray<IrisKind | Kind> = [
        IrisKind.TYPE_DEFINITION,
        IrisKind.VARIANT_DEFINITION,
      ];
      return kinds.includes(parentNode.kind)
        ? DirectiveLocation.INPUT_FIELD_DEFINITION
        : DirectiveLocation.ARGUMENT_DEFINITION;
    }
    default:
      invariant(false, 'Unexpected kind: ' + inspect(appliedTo.kind));
  }
}
