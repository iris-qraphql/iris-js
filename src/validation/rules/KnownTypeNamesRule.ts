import { irisNodeError } from '../../error';
import type { ASTNode } from '../../types/ast';
import {
  isTypeDefinitionNode,
  isTypeSystemDefinitionNode,
} from '../../types/ast';
import { scalarNames } from '../../types/definition';
import type { ASTVisitor } from '../../types/visitor';
import { didYouMean, suggestionList } from '../../utils/legacy';

import type { SDLValidationContext } from '../ValidationContext';

/**
 * Known type names
 *
 * A GraphQL document is only valid if referenced types (specifically
 * variable definitions and fragment conditions) are defined by the type schema.
 *
 * See https://spec.graphql.org/draft/#sec-Fragment-Spread-Type-Existence
 */
export function KnownTypeNamesRule(context: SDLValidationContext): ASTVisitor {
  const definedTypes = Object.create(null);
  for (const def of context.getDocument().definitions) {
    if (isTypeDefinitionNode(def)) {
      definedTypes[def.name.value] = true;
    }
  }

  const typeNames = [...Object.keys(definedTypes)];

  return {
    NamedType(node, _1, parent, _2, ancestors) {
      const typeName = node.name.value;
      if (!definedTypes[typeName]) {
        const definitionNode = ancestors[2] ?? parent;
        const isSDL = definitionNode != null && isSDLNode(definitionNode);
        if (isSDL && scalarNames.includes(typeName)) {
          return;
        }

        const suggestedTypes = suggestionList(
          typeName,
          isSDL ? scalarNames.concat(typeNames) : typeNames,
        );
        context.reportError(
          irisNodeError(
            `Unknown type "${typeName}".` + didYouMean(suggestedTypes),
            node,
          ),
        );
      }
    },
    VariantDefinition(node, _1, parent, _2, ancestors) {
      if (node.fields) {
        return undefined;
      }

      const typeName = node.name.value;
      if (!definedTypes[typeName]) {
        const definitionNode = ancestors[2] ?? parent;
        const isSDL = definitionNode != null && isSDLNode(definitionNode);
        if (isSDL && scalarNames.includes(typeName)) {
          return;
        }

        const suggestedTypes = suggestionList(
          typeName,
          isSDL ? scalarNames.concat(typeNames) : typeNames,
        );
        context.reportError(
          irisNodeError(
            `Unknown type "${typeName}".` + didYouMean(suggestedTypes),
            node,
          ),
        );
      }
    },
  };
}

function isSDLNode(value: ASTNode | ReadonlyArray<ASTNode>): boolean {
  return 'kind' in value && isTypeSystemDefinitionNode(value);
}
