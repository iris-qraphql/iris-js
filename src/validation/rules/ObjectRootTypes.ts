import { irisError } from '../../error';
import { isTypeVariantNode } from '../../types/ast';
import type { ASTVisitor } from '../../types/visitor';

import type { IrisValidationContext } from '../ValidationContext';

export const ObjectRootTypes = (ctx: IrisValidationContext): ASTVisitor => ({
  TypeDefinition(type) {
    const typeName = type.name.value;
    if (['Query', 'Mutation', 'Subscription'].includes(typeName)) {
      if (!(type.role === 'resolver' && isTypeVariantNode(type))) {
        const optionalMessage = typeName === 'Query' ? '' : ' if provided';
        ctx.reportError(
          irisError(
            `${typeName} root type must be Object type${optionalMessage}, it cannot be ${typeName}.`,
            { nodes: type.name },
          ),
        );
      }
    }
  },
});
