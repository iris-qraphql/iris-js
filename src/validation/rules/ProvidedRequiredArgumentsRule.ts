import type { ASTVisitor } from '../../language/visitor';

import { irisError } from '../../error';
import { print } from '../../printing/printer';
import type { ArgumentDefinitionNode } from '../../types/ast';
import type { IrisArgument } from '../../types/definition';
import { isRequiredArgument, isType } from '../../types/definition';
import { specifiedDirectives } from '../../types/directives';
import { IrisKind } from '../../types/kinds';
import { inspect } from '../../utils/legacy';
import type { ObjMap } from '../../utils/ObjMap';
import { keyMap } from '../../utils/ObjMap';

import type { SDLValidationContext } from '../ValidationContext';

/**
 * @internal
 */
export function ProvidedRequiredArgumentsOnDirectivesRule(
  context: SDLValidationContext,
): ASTVisitor {
  const requiredArgsMap: ObjMap<ObjMap<IrisArgument | ArgumentDefinitionNode>> =
    Object.create(null);

  const schema = context.getSchema();
  const definedDirectives = schema?.directives ?? specifiedDirectives;
  for (const directive of definedDirectives) {
    requiredArgsMap[directive.name] = keyMap(
      directive.args.filter(isRequiredArgument),
      (arg) => arg.name,
    );
  }

  const astDefinitions = context.getDocument().definitions;
  for (const def of astDefinitions) {
    if (def.kind === IrisKind.DIRECTIVE_DEFINITION) {
      const argNodes = def.arguments ?? [];

      requiredArgsMap[def.name.value] = keyMap(
        argNodes.filter(isRequiredArgumentNode),
        (arg) => arg.name.value,
      );
    }
  }

  return {
    Directive: {
      // Validate on leave to allow for deeper errors to appear first.
      leave(directiveNode) {
        const directiveName = directiveNode.name.value;
        const requiredArgs = requiredArgsMap[directiveName];
        if (requiredArgs) {
          const argNodes = directiveNode.arguments ?? [];
          const argNodeMap = new Set(argNodes.map((arg) => arg.name.value));
          for (const [argName, argDef] of Object.entries(requiredArgs)) {
            if (!argNodeMap.has(argName)) {
              const argType = isType(argDef.type)
                ? inspect(argDef.type)
                : print(argDef.type);
              context.reportError(
                irisError(
                  `Directive "@${directiveName}" argument "${argName}" of type "${argType}" is required, but it was not provided.`,
                  { node: directiveNode },
                ),
              );
            }
          }
        }
      },
    },
  };
}

function isRequiredArgumentNode(arg: ArgumentDefinitionNode): boolean {
  return arg.type.kind !== IrisKind.MAYBE_TYPE && arg.defaultValue == null;
}
