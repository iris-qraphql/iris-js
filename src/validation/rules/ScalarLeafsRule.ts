import { inspect } from '../../jsutils/inspect';

import type { FieldNode } from '../../language/ast';
import type { ASTVisitor } from '../../language/visitor';

import { getNamedType, isDataType } from '../../type/definition';

import { GraphQLError } from '../../error';

import type { ValidationContext } from '../ValidationContext';

/**
 * Scalar leafs
 *
 * A GraphQL document is valid only if all leaf fields are of scalar or data types.
 */
export function ScalarLeafsRule(context: ValidationContext): ASTVisitor {
  return {
    Field(node: FieldNode) {
      const type = context.getType();
      const selectionSet = node.selectionSet;
      if (type) {
        if (isDataType(getNamedType(type))) {
          if (selectionSet) {
            const fieldName = node.name.value;
            const typeStr = inspect(type);
            context.reportError(
              new GraphQLError(
                `Field "${fieldName}" must not have a selection since type "${typeStr}" has no subfields.`,
                selectionSet,
              ),
            );
          }
        } else if (!selectionSet) {
          const fieldName = node.name.value;
          const typeStr = inspect(type);
          context.reportError(
            new GraphQLError(
              `Field "${fieldName}" of type "${typeStr}" must have a selection of subfields. Did you mean "${fieldName} { ... }"?`,
              node,
            ),
          );
        }
      }
    },
  };
}
