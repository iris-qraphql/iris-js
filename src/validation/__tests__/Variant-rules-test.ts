import { getSDLValidationErrors } from '../../utils/toJSONDeep';

import { IncludeOnlyVariantTypes } from '../rules/IncludeOnlyVariantTypes';

const expectSDLErrors = (sdlStr: string) =>
  expect(getSDLValidationErrors(IncludeOnlyVariantTypes, sdlStr));

describe('VariantType Rules', () => {
  it('rejects a Union type with non-Object members types', () => {
    expectSDLErrors(`
      data A = {
        field: String
      }

      data BadUnion = A | String | Int
    `).toMatchSnapshot();
  });

  it("can't build recursive Union", () => {
    expectSDLErrors(`
    data Hello = Hello
  `).toMatchSnapshot();
  });
});
