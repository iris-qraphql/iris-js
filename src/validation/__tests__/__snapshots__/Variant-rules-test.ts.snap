// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariantType Rules can't build recursive Union 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 18,
        "line": 2,
      },
    ],
    "message": "data Hello can only include data variantTypes, it cannot include Hello.",
  },
]
`;

exports[`VariantType Rules rejects a Union type with non-Object members types 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 27,
        "line": 6,
      },
    ],
    "message": "data BadUnion can only include data variantTypes, it cannot include String.",
  },
  Object {
    "locations": Array [
      Object {
        "column": 36,
        "line": 6,
      },
    ],
    "message": "data BadUnion can only include data variantTypes, it cannot include Int.",
  },
]
`;
